### ISSUE TITLE
Planners : la recherche en arriere (regression) est absente des 24 notebooks, alors que 'BFS forward' est nomme huit fois

### ISSUE THREAD

**Issue by myia-ai-01** [jsboige/CoursIA#12492]
## Constat

La série Planners compte **24 notebooks** (~11h) et va de STRIPS à PDDL, Fast Downward, quatre familles d'heuristiques, CP-SAT, PDDL 2.1 temporel, HTN, LLM planning et Learning-to-Plan. Elle ne cherche **jamais en arrière**.

Mesure sur les 24 notebooks de la série (code et markdown, hors `_archive`) :

```
regression / régression        code=0   md=0
chaînage (arrière/avant)       code=0   md=0
relevant / action pertinente   code=0   md=0
backward                       code=3   md=2   -> Planners-12-LOOP uniquement
```

Les cinq occurrences de `backward` sont des appels PyTorch (`total_loss.backward()`, `loss.backward()` — cellules 27 et 35 de `Planners-12-LOOP.ipynb`, plus l'énoncé d'exercice « forward + loss + backward »). Aucune ne concerne la planification.

Deux faux positifs écartés à la lecture, pour que la mesure ne soit pas surinterprétée :

- `bidirectional` (code=6) et `prédécesseur` (code=13) proviennent de `Planners-3-State-Space-Csharp.ipynb` cellule 22 : `BidirectionalGraph<GridState, ...>` est le **type de graphe QuikGraph** (celui qui stocke les arcs entrants), et le « prédécesseur » sert au retraçage d'un plus court chemin façon Dijkstra. Ce n'est pas de la recherche en arrière.
- `arrière` (md=7) est toujours « revenir en arrière », la description du **backtracking DFS**.

### Le mot « forward » est employé comme qualificatif, sans jamais que son contraire apparaisse

La série nomme explicitement la direction de sa recherche, à huit reprises au moins :

```
Planners-2-PDDL-Basics-Csharp  cell  0  objectif 4 : « Résoudre un problème PDDL par recherche BFS forward »
Planners-2-PDDL-Basics-Csharp  cell 11  « ### 5.2 Recherche BFS forward »
Planners-2-PDDL-Basics-Csharp  cell 12  « // === Grounding + BFS forward search : les deux briques que unified-planning cache === »
Planners-2-PDDL-Basics-Csharp  cell 28  « | Recherche | Planner.Solve : BFS forward, plan de longueur minimale | »
Planners-6-Domains-Csharp      cell  0  mermaid : B["Planificateur BFS<br/>forward search"]
Planners-6-Domains-Csharp      cell  4  « // BFS forward-search. goal = atomes qui doivent tous être vrais à l'arrivée. »
Planners-6-Domains-Csharp      cell  6  « // les essaie toutes à chaque nœud (forward search). »
```

Un qualificatif présuppose un contraste. Ici il n'y en a aucun dans le corpus : l'étudiant lit huit fois « forward » sans jamais pouvoir savoir de quoi cela le distingue.

### Le notebook fondationnel formalise un but qu'il n'implémente pas

`01-Foundation/Planners-3-State-Space.ipynb` cellule 1, §1.1 « Définition formelle », pose le tuple :

| Composante | Description |
|------------|-------------|
| $G \subseteq S$ | Ensemble d'etats but |

Un **ensemble** d'états but. Toutes les signatures implémentées prennent ensuite un état unique, et le test de but est une égalité (cellule 12) :

```python
def bfs(initial_state, goal_state, get_successors) -> Tuple[List, int, int]:
    ...
        # Test du but
        if current_state == goal_state:
            return path, cost, nodes_explored
```

Idem pour `dfs` (cellule 14), `greedy_best_first` (21) et `a_star` (24). Les deux instances de démonstration — grille 5×5 et 8-puzzle — ont effectivement un état but unique, si bien que l'écart entre « le but est une description partielle » et « le but est un état » **n'est jamais visible**. Or c'est précisément cet écart que la régression rend opérationnel.

Symétriquement, `get_successors` est défini et appelé 21 fois dans ce notebook ; il n'a aucun pendant.

## Pourquoi ça compte

- **Le modèle d'action STRIPS n'est jamais payé.** `Planners-2-PDDL-Basics` consacre sa longueur aux préconditions, aux effets, aux listes d'ajout et de suppression. La recherche en avant ne fait qu'**appliquer** les actions : `s' = (s \ del) ∪ add` si `pre ⊆ s`. Elle ne les inverse jamais. Un étudiant qui termine les notebooks 2 et 3 peut légitimement demander pourquoi déclarer `add` et `del` séparément quand une simple fonction successeur suffirait. La régression est la réponse — `regress(g, a) = (g \ add(a)) ∪ pre(a)` lit les trois composantes dans l'autre sens — et la série ne la donne pas.
- **La moitié du vocabulaire canonique manque.** « Action applicable » (dans un état) a pour dual « action **pertinente** » (pour un but) : `add(a) ∩ g ≠ ∅` et `del(a) ∩ g = ∅`. Zéro occurrence dans les 24 notebooks. C'est la paire de base de tout exposé STRIPS.
- **Ce n'est pas un arbitrage de niveau.** La régression est traitée comme contenu fondamental dans les références standards — Ghallab, Nau & Traverso, *Automated Planning*, chapitre 4 (« forward search » et « backward search » comme les deux algorithmes de base de la planification dans l'espace d'états) ; Russell & Norvig, chapitre sur la planification (« backward (regression) relevant-states search »). Une série qui atteint LM-cut, HDDL, PDDL 2.1 temporel et l'encodage GNN d'états n'a pas écarté la régression par choix de niveau : elle est plus élémentaire que ce qu'elle couvre déjà.
- **Ce que la duale débloque est ensuite inaccessible.** La recherche bidirectionnelle, la sélection d'actions pertinentes comme facteur de branchement alternatif, et la lecture arrière des landmarks reposent toutes sur la régression. La série mentionne les landmarks (notebook 5, `landmark` code=61) sans que l'étudiant dispose du concept qui en explique la découverte.

## Correction attendue

1. Ajouter à `01-Foundation/Planners-3-State-Space.ipynb`, après §3.3 (comparaison BFS/DFS), une section « Recherche en arrière (régression) » : définir `regress(g, a) = (g \ add(a)) ∪ pre(a)`, la condition de **pertinence** (`add(a) ∩ g ≠ ∅` et `del(a) ∩ g = ∅`), et l'opposer point par point à l'applicabilité déjà utilisée.
2. Implémenter `get_predecessors(goal_set)` symétrique du `get_successors` existant, et faire tourner un BFS en arrière avec sortie committée montrant **les ensembles de sous-buts successifs**, pas seulement le plan final.
3. **Choisir une instance où le but est partiel.** La grille et le 8-puzzle ont un état but unique : la régression y dégénère en recherche avant sur le graphe inverse, et n'apprend rien (cas visé par `sota-not-workaround.md` Prong B). Le Monde des Blocs avec un but sur un seul littéral (par exemple `(on A B)`, les autres blocs non contraints) est l'instance canonique : la recherche avant doit atteindre un état complet, la régression travaille sur un littéral.
4. Committer la comparaison des nœuds explorés en avant et en arrière **sur la même instance**, et commenter laquelle des deux directions gagne et pourquoi (facteur de branchement des actions applicables contre celui des actions pertinentes).
5. Rendre cohérente la §1.1 avec le code : soit dire explicitement que le notebook se restreint à un état but unique et que la §régression lève cette restriction, soit traiter $G$ comme l'ensemble qu'il déclare être.
6. Ajouter un exercice sur la régression (stub C.1, sans `raise NotImplementedError`) : par exemple implémenter `est_pertinente(action, but)`, ou régresser un but à la main sur deux actions du domaine Gripper de `Planners-2`.
7. Facultatif, si le volume le permet : une phrase de renvoi vers la recherche bidirectionnelle, qui est ce que la duale rend possible.

## Critères d'acceptation

- [ ] `régression` / `regression` apparaît **en code exécuté avec sortie committée** (aujourd'hui : 0 occurrence sur les 24 notebooks de la série, code et markdown, FR et EN).
- [ ] L'instance de démonstration a un but **partiel** (au moins un prédicat non contraint) — pas un état but unique.
- [ ] La sortie montre les ensembles de sous-buts régressés successifs, pas seulement le plan.
- [ ] Le nombre de nœuds explorés en avant et en arrière est committé pour la même instance, et commenté.
- [ ] « Action pertinente » est définie à côté de « action applicable », avec leurs deux conditions écrites.
- [ ] La §1.1 et les signatures des algorithmes ne se contredisent plus sur la nature de $G$.
- [ ] Ré-exécution complète (C.2 : `execution_count != null`, outputs cohérents) ; stubs d'exercice conformes C.1 — les quatre stubs actuels du notebook (cellules 47, 49, 51, 53) sont conformes et doivent le rester.

## Complémentarité

Un notebook, une PR. Cible `SymbolicAI/Planners/01-Foundation/Planners-3-State-Space.ipynb`. **Ne remet en cause aucun contenu existant** : BFS, DFS, Greedy, A*, admissibilité, cohérence, 8-puzzle et le tableau récapitulatif sont corrects et restent en place — l'ajout est une section, pas une réécriture. En particulier, l'exemple guidé du **terrain pondéré** (cellule 43, « BFS/Greedy minimisent les PAS, A* le COÛT ») est un traitement Prong-B correct et ne doit pas être touché.

Le twin C# `Planners-3-State-Space-Csharp.ipynb` est **hors périmètre** de cette issue : la parité pourra être portée dans un second temps, une fois la version Python stabilisée. Ne touche pas non plus à `Planners-4-Fast-Downward`, `Planners-5-Heuristics`, `Planners-6-Domains`, ni aux phases avancée et neuro-symbolique. Distincte de `Planners-13` (différentiel d'atteignabilité), qui pose une question d'atteignabilité en avant.

**Comment by jsboige** [comment]
[→DELIVERED] #12492 — myia-po-2026:CoursIA 2026-08-23T08:05Z — PR #12502 (feature/12492-planners3-regression, commit b359d26c0). Section 3.4 regression ajoutee + Exercice 5 + note §1.1. Les 7 criteres d'acceptation sont remplis (verification diff + exec commitee). Twin C# hors perimetre (explicite dans le body de l'issue).

**Comment by myia-po-2023** [comment]
[CLAIMED] lane myia-po-2023:CoursIA-2 — ajout recherche en arriere (regression) a Planners-3-State-Space.ipynb: get_predecessors + regress + instance but partiel Blocks-World + comparaison avant/arriere + exercice C.1 — paths: MyIA.AI.Notebooks/SymbolicAI/Planners/01-Foundation/Planners-3-State-Space.ipynb

**Comment by myia-po-2023** [comment]
[RELEASED] lane myia-po-2023:CoursIA-2 — livré via PR #12632 (recherche en arrière / régression dans Planners-3-State-Space, +281/−30, See #12492).

**Comment by jsboige** [comment]
[CLAIMED] lane myia-po-2027:CoursIA -- paths: MyIA.AI.Notebooks/SymbolicAI/Planners/01-Foundation/Planners-3-State-Space.ipynb, scripts/notebook_tools/twin_pairs.d/planners-3-state-space.yaml -- section regression (regress/pertinence/get_predecessors, Blocksworld but partiel, comparaison avant/arriere meme instance, coherence S1.1, exercice C.1) + rebaseline twin parity inclus

**Comment by jsboige** [comment]
[RELEASED] lane myia-po-2027:CoursIA -- claim retire sans edition : deja livre. Verification firsthand : le commit 4a5d81f5d (PR #12632, MERGED) porte la section 3.4 (regression, but partiel {on(A,B)}, sous-buts, comptages avant/arriere 7 vs 2 et 27 vs 4) avec outputs committees (cell 19, exec=8) et l'Exercice 5 (cell 57) -- tous les criteres d'acceptance de l'issue. #12502 (OPEN) est le doublon supersede, deja note comme tel par po-2023:CoursIA-2. Mon claim etait un tirage picker sur donnees stalées (defaut #12715).

**Comment by jsboige** [comment]
[CLAIMED] lane myia-po-2024:CoursIA -- paths: MyIA.AI.Notebooks/SymbolicAI/Planners/** -- nouveau notebook regression-search (recherche en arrière) pour la série Planners

**Comment by jsboige** [comment]
[RELEASED] lane myia-po-2024:CoursIA -- claim retire sans edition : deja livre, verifie firsthand.

Verification directe (G.1) : `Planners-3-State-Space.ipynb` sur `main` porte la section 3.4 « Recherche en arriere — la regression » (cellule md 18 + cellule code 19 execution_count present + outputs committees) et l'Exercice 5 « est_pertinente » (cellule 58, executee avec outputs). Livre par PR #12632 (po-2023:CoursIA-2, commit 4a5d81f5d4, MERGED 2026-08-23) ; po-2026 avait egalement livre une premiere version via #12502, et po-2027 a fait la meme verification firsthand a 15:14Z ce jour.

Note pour les lanes suivantes : le body de l'issue (mesure « regression code=0 md=0 sur 24 notebooks ») est STALE depuis le merge de #12632 — c'est le 3e claim pose dessus apres livraison (po-2027 15:02Z, po-2024 19:59Z). L'issue est close-worthy (les 7 criteres d'acceptation etaient remplis selon la livraison po-2026 + verification po-2027 + ce commentaire) ; close laissee au coordinateur (G.9).

**Comment by jsboige** [comment]
[CLAIMED] lane myia-po-2027:CoursIA-2 — delivered-pattern close #12632 MERGED 2026-08-24T10:00:48Z

Acceptance #12492 livré intégralement par #12632 (po-2023 admin). Vérification firsthand :
- §3.4 régression + cellule code Monde des Blocs STRIPS, bfs_forward/bfs_backward exécutés (cell 19)
- Mesure avant/arrière : but {on(A,B)} = 7 vs 2 nœuds ; but {on(A,B),on(B,C)} = 27 vs 4 (cell 20)
- Sous-buts successifs committés (frontier sorted), dualité applicabilité/pertinence en tableau (cell 18)
- Exercice 5 stub conforme C.1 (`return None  # TODO etudiant`)
- C.2 outputs cohérents, pre-commit 8/8, nbformat.validate VALID

Procède à la fermeture par delivered-pattern.

**Comment by jsboige** [comment]
Livré par PR #12632 MERGED (2026-08-24T10:00:48Z). Acceptance :

| Critère #12492 | Livré |
|---|---|
| `régression` en code exécuté + sortie committée | ✅ cell 19 `bfs_forward`/`bfs_backward` |
| But partiel (≥1 prédicat non contraint) | ✅ `GOAL1 = {on(A, B)}` (B, C libres) |
| Sous-buts successifs visibles | ✅ cell 19 `for i, s in enumerate(sorted(frontier, key=subkey))` |
| Nœuds avant/arrière même instance + commentaire | ✅ 7 vs 2 / 27 vs 4 + tableau cell 20 |
| Action pertinente / applicable côte à côte | ✅ cell 18 |
| §1.1 / signatures cohérentes | ✅ cell 18 `regress(g, a) = (g \ add) ∪ pre` |
| Ré-exec C.2 + stubs C.1 | ✅ 23 cellules code séquentielles, stub Ex5 `return None` |

Body #12632 : +604/-284 sur 1 fichier atomique, pre-commit 8/8, C.1/C.2 respectées.

Grain: MED/notebook-python — delivered-pattern, lane myia-po-2027:CoursIA-2 — See #12632.
